Blueprint: add support for custom files and directories in `/etc`

thozza opened the pull request:
Tl;DR; this PR adds the ability to create custom files and directories under `/etc` in the image by specifying them as a BP customization. The functionality is supported for all non-ostree images and for ostree commit and container images.

| | Feature description | Support in this PR | Comment |
| --- | --- | :---: | --- |
| Directories customization |Create new | :white_check_mark: | <ul><li>possible to specify that parent directories should be created as needed</li><li>path policy applies</li></ul> |
| | Modify existing | :x: | <ul><li>Modifying ownership and permissions of existing directories is not supported and will result in failure during image build.</li><li>One can specify only plain custom directory (without explicit ownership and permissions) to ensure that the directory exist (e.g. due to bing a parent of a custom file). In such case, the directory will be created with default permissions and ownership if it does not exist in the image or it will be a noop if it exists.</li></ul>|
| | Delete existing | :x:| |
| |Set user / group ownership| :white_check_mark: | <ul><li>possible to use using user / group name or UID / GID</li><li>possible to set to user / group provided as BP customization</li></ul> |
| | Set mode | :white_check_mark: | <ul><li>possible by providing mode in octal format (such as "0755")</li></ul> |
| | Set additional attributes | :white_check_mark: | <ul><li>such as xattr, custom SELinux labels, etc.</li></ul> |
| Files customization | Create new | :white_check_mark: | <ul><li>path policy applies</li><li>parent directories must exist</li></ul> |
| | Modify existing | :white_check_mark: | <ul><li>all supported file attributes can be set on existing files and their content overridden</li><li>path policy applies</li></ul> |
| | Delete existing | :x: | |
| | Set user / group ownership | :white_check_mark: | <ul><li>possible to use using user / group name or UID / GID</li><li>possible to set to user / group provided as BP customization</li></ul> |
| | Set mode | :white_check_mark: | <ul><li>possible by providing mode in octal format (such as "0755")</li></ul> |
| | Set additional attributes | :x: | <ul><li>such as xattr, custom SELinux labels, etc. </li></ul> |
| | Plain text content | :white_check_mark: | |
| |Binary (base64) content | :x: | <ul><li>would be very easy to add</li></ul>|
| Support in distros | all | :white_check_mark: | |
| Support in image types | non-ostree image types | :white_check_mark: | <ul><li>such as qcow2, cloud images, etc.</li></ul> |
| |<ul><li>edge-commit</li><li>edge-container</li></ul>| :white_check_mark: | <ul><li>files and dir are baked into the commit</li></ul> |
| | <ul><li>edge-installer</li><li>edge-raw-image</li><li>edge-simplified-installer</li></ul> | :x: | <ul><li>not as a deploy-time customization</li></ul>|
| Common | SELinux labels | :grey_question: | <ul><li>Custom directories and files will have the default label applied as specified by the used SELinux policy</li></ul>|
| | Weldr API support | :white_check_mark: | |
| | Cloud API support | :white_check_mark: | |
| | Limits | ❔ | <ul><li>No limit is applied on the number of custom directories and files that can be created</li><li>No limit is applied on the size of custom files</li><li>No assurance that the custom files can fit in the image or partition if it is too small</li></ul>|

Most notable changes:
- Add internal representation for filesystem nodes as `fsnode` package
- Add support for `org.osbuild.chown` stage
- Add helper functions for generating osbuild stages from `fsnode` types with the aim to create them in the image.
- Extend Blueprint customizations with Directories and Files customizations and implement their validation.
- Extend Cloud API to accept Directories and Files customizations and validate them.
- Extend distro definitions and pipeline implementations to handle custom Directories and Files.
- Extend image test manifests with the new customizations and regenerate them.
- Extend the OSTree tests (`OSTree` and `New OSTree`) to use the custom Directories and FIles feature and test its effect on the resulting deployed image.

Details and some implementation choices are described in commit messages.

TODO:

- [x] Determine desired custom directories policy (which directories to protect?)
  - Dirs can be created only under `/etc` and existing dirs can't be modified. 
- [x] Determine desired custom files policy (which files to protect?)
  - Keep it minimal (fstab, groups, passwd, shadow)

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
  - [x] unit tests
  - [x] Weldr API tests (tested via ostree tests, which use it)
  - [x] Cloud API tests
  - [x] ostree functional tests (ostree / ostree-ng)
- [x] adequate documentation informing people about the change such as
  - [x] https://github.com/osbuild/guides/pull/103

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`

Information regarding our GitLab pipeline (Schutzbot):

CI will not be ran automatically if WIP label is applied or the PR is in DRAFT state, instead only
a link will be provided to the pipeline which can then be triggered manually if desired. To run the
CI automatically either switch the PR to ready or apply WIP+test label.

Outside contributors need manual approval from one of the osbuild-composer maintainers.

Schutzbot will only be triggered if all Tests jobs in GitHub workflow succeed.
-->

henrywang commented:
@yih-redhat Could you please take a look at this PR and add test here? Thanks.

yih-redhat commented:
> @yih-redhat Could you please take a look at this PR and add test here? Thanks.

Tomas will add the test code, we can extend the test if necessary once it's ready.

lavocatt commented on internal/osbuild/copy_stage.go:
```diff
@@ -39,6 +39,10 @@ func NewCopyStageSimple(options *CopyStageOptions, inputs Inputs) *Stage {
 	}
 }
 
+type CopyStageFilesInputs map[string]*FilesInput
```
There's a `TODO: commit message` for this commit :smile:

thozza replied:
Can you post the commit SHA? Github is not good at showing that with a review...

thozza replied:
Got it... https://github.com/osbuild/osbuild-composer/pull/3281/commits/b354e86bbabcbb865a08612401affe12418d941d I'll add a message then :D

thozza replied:
I'll wait for other reviews before pushing the fixed commit

lavocatt commented on internal/distro/rhel7/images.go:
```diff
@@ -141,6 +142,21 @@ func osCustomizations(
 	}
 
+	var err error
+	osc.Directories, err = blueprint.DirectoryCustomizationsToFsNodeDirectories(c.GetDirectories())
```
Can you generalize this chunk of code to someting like ?
```
osc.Directoriies, osc.Files, err = bluepring.somename(...)
```
To avoid copy pasting this 3 times.

thozza replied:
I admit that his is a bit verbose, but this is unfortunately the way how error handling works in Go. I do not consider this to be an issue to be honest.

I do not want to unnecessarily tie Files and Directories customization together when not needed. These are two different customizations. We do not do what you are suggesting for any other customizations (e.g. users / groups).

Lastly, the fact that this is copied three times in each distro is a result of how we define distros, not of the fact that there is a separate function for each customization, as there already is for those existing ones. Instead, we should aim for deduplicating `distro/<distro>/images.go` where it makes sense, because these are very similar. However this was not the goal of the rework that @achilleas-k did in all distros, but it is AFAIK something that we would like to eventually look at.

lavocatt commented on internal/distro/fedora/images.go:
```diff
@@ -385,6 +385,16 @@ func iotRawImage(workload workload.Workload,
 	img.Groups = users.GroupsFromBP(customizations.GetGroups())
 
+	var err error
+	img.Directories, err = blueprint.DirectoryCustomizationsToFsNodeDirectories(customizations.GetDirectories())
```
Same here for the generalization

thozza replied:
Commented in the Fedora distro. TL;DR; I'm not in favor of generalizing this.

lavocatt commented on internal/distro/rhel8/images.go:
```diff
@@ -466,6 +466,17 @@ func edgeSimplifiedInstallerImage(workload workload.Workload,
 
 	rawImg.KernelOptionsAppend = []string{"modprobe.blacklist=vc4"}
+
+	var err error
```
Same here for the generalization

thozza replied:
Commented in the Fedora distro. TL;DR; I'm not in favor of generalizing this.

lavocatt reviewed (commented):
Very nice one! I have a few comments/questions below.

I'm wondering also, has the `etc` customization been ported to every test cases that supports it ? If so, should we also add some tests without `etc` customization for the completeness of the testing scenarios.

thozza commented:
> Very nice one! I have a few comments/questions below.

Thanks for your review!

> I'm wondering also, has the `etc` customization been ported to every test cases that supports it ? If so, should we also add some tests without `etc` customization for the completeness of the testing scenarios.

So just to clear any confusion, the PR does not add `etc` customization, but **two** new customizations for Files and Directories. That may be helpful when thinking about why these are not handled together unless completely necessary. There is only one case where this is needed, to check that someone didn't specify a custom file, which is a parent of a custom directory. We could just fail during build, because directories are created first, but I decided to do the validation in composer to fail fast.

As for the test cases:

- I didn't want to substantially grow our functional test jobs, because I don't see any value in that (than paying more to AWS). I considered adding a separate test case, like the filesystem customization has. After a discussion with @ondrejbudai and @croissanne, I decided to just extend the API test case.
- I extended test cases which already were testing some customizations (mostly API and ostree tests) to use it. Note that ostree tests do not use it on all distros, only on Fedora all, CentOS Stream 8/9 and rhel-9.2 and rhel-8.8.
- One can consider the vanilla image test manifests without any customizations to cover that case when the customization is not used (IOW that nothing gets added to the manifest). I don't think that we can do any additional testing of "not using the new customizations" that would extend the coverage.

croissanne commented on internal/distro/fedora/images.go:
```diff
@@ -129,6 +129,21 @@ func osCustomizations(
+	if err != nil {
+		// In theory this should never happen, because the blueprint directory customizations
+		// should have been validated before this point.
+		panic(fmt.Sprintf("failed to convert directory customizations to fs node directories: %v", err))
```
I guess it's fine to panic because this would be a programming error. But why not propagate the error? All functions which invoke osCustomizations can return an error (but none do atm). Maybe the panic would need to happen higher up? idk, not a blocker any way, just wondering.

thozza replied:
I didn't return the error, because as you mentioned, the `osCustomizations()` does not return any. Personally I prefer to return errors instead of panicking. These customizations were the only ones that could return error while converting them. I think that we may end up returning errors also for other customizations if we start validating them.

I would be happy to rework this to return an error... 😉

croissanne replied:
Maybe in a followup though :D

croissanne reviewed (commented):
lgtm other than the TODO commit msg :)

thozza commented:
Note that the files customization would currently follow symlinks if the file path already exists in the image and it is a symlink to a file. I'll address this in a follow-up PR once https://github.com/osbuild/osbuild/pull/1241 is merged.

mrguitar commented:
@thozza looks great! Thanks for tweaking the commit vs installer deployment. Two small questions:

1) if a user creates a unit file, will the ordering in osbuild allow the [customization.services] to enable/disable the new units?
2) We're pasting in plain text in between quotes. I haven't gotten my hands on this, nor am I a TOML person, so apologies if this is a dumb question. Will users have to escape quotes that are contained inside the pasted text or anything like that?

achilleas-k commented on internal/disk/disk.go:
```diff
@@ -184,19 +168,3 @@ func NewVolIDFromRand(r *rand.Rand) string {
 	return hex.EncodeToString(volid)
 }
-
-func CheckMountpoints(mountpoints []blueprint.FilesystemCustomization, mountpointAllowList *pathpolicy.PathPolicies) error {
```
I think this fits better in the new `pathpolicy` package instead of `blueprint` and should have probably moved with the rest of the policy parts.  This shouldn't create any import issues since the `pathpolicy` is imported in `blueprint` only for these checks.

Since the policies are created and managed in the `pathpolicy` package, it seems more natural that the code to use them for the main purpose (policy checking) would be in that same package instead of where the values being checked are defined (`blueprint`).

thozza replied:
While I agree that policies are currently created and managed in the `pathpolicy` package, It could be applied on any object with paths. So in theory it could be applied also on a slice of `disk.Filesystem` or any other.

In the same sense as we do not refer to `osbuild` package structures in the internal API-agnostic code, I do believe that it is also good to not refer to the `blueprint` package structures outside of it unless necessary to convert them. I get that `blueprint` is kind of API-agnostic concept in composer so comparing it to `osbuild` package is maybe a stretch.

Lastly, my thinking was also to keep all helper functions for `blueprint` structures in that package, so those that transform them to the "internal API" structures, but also those that are doing come validation  / (policy) checks on them.

But I do not feel very strongly about keeping this in the `blueprint` package.

achilleas-k replied:
I see your point.  Let's keep it the way it is now then, in the `blueprint` package.
I'm coming around to your way of thinking of having the "internal" structures never import or refer to neither `blueprint` nor `osbuild`.

achilleas-k commented on internal/disk/disk.go:
```diff
@@ -53,19 +50,6 @@ const (
-	"/data": {},
-	"/home": {},
-	"/tmp":  {},
-})
```
Can this change (moving the `MountpointPolicies`) be merged with the commit when the package split happened?  Not important though.

thozza replied:
Probably yes, I just did the split really in steps to not mess things up too much. Thus moving things around in multiple commits.

achilleas-k replied:
No worries then.

achilleas-k commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,284 @@
+// validateModeString checks that the given string is a valid mode octal number
+func validateModeString(mode string) error {
+	// check that the mode string matches the octal format regular expression
+	if regexp.MustCompile(`^0[0-7]{3}$`).MatchString(mode) {
+		return nil
+	}
```
Are we sure we want to restrict it to having a leading `0`?  I suspect a lot of users will get tripped up by this from the general habit of `chmod`ing with just the three digits for convenience.

thozza replied:
I think I could change the regex to make the leading `0` optional. I agree that people are probably more used to write permissions without the leading zero.

achilleas-k reviewed (commented):
Partial review and have to stop for the day.

Reached a69ea5f89cb6cb54737101552fe89b552dd41175

teg commented:
Only high level review, but looks great!

thozza commented on internal/fsnode/file.go:
```diff
@@ -0,0 +1,36 @@
+
+// NewFile creates a new file with the given path, data, mode, user and group.
+// user and group can be either a string (user name/group name), an int64 (UID/GID) or nil.
+func NewFile(path string, data []byte, mode *os.FileMode, user interface{}, group interface{}) (*File, error) {
```
I'm not opposed to it. I considered this during the implementation and went with the current version due to the fact, that one would probably always want to specify `data` for a file, but not necessarily the `mode`, `user`, etc. However, now it feels as if I was thinking about this more in the `Python` terms in the sense that in `Go` you can't specify default values for arguments and thus you must always provide all of them. So having the `data` as the second argument does not help much.

achilleas-k replied:
I don't mind either way.  :)

thozza commented on internal/distro/rhel9/imagetype.go:
```diff
@@ -360,7 +360,7 @@ func (t *imageType) checkOptions(customizations *blueprint.Customizations, optio
 			}
 
 			// ignition is optional, we might be using FDO
-			if customizations.Ignition.HasIgnition() {
```
Thanks, will check it out... 🤔

thozza commented on internal/fsnode/fsnode.go:
```diff
@@ -0,0 +1,132 @@
+	case nil:
+		// user is not set
+	default:
+		return fmt.Errorf("user must be either a string or an int64, got %T", user)
```
It can be both, therefore we can not always `panic`...

1. If one defines a new `fsnode` node type in the code with values which do not validate, then it would be a programming error.
2. The validation is later (ab)used when validating user input from the Blueprint customization (`blueprint` package and `cloudapi` package) and returning an error to the user if they provided invalid input. This was my way of not having to implement the validation in 3 places, but only in one place.

achilleas-k replied:
Right.  Sounds good to me.

thozza commented:
> @thozza looks great! Thanks for tweaking the commit vs installer deployment. Two small questions:
> 
> 1. if a user creates a unit file, will the ordering in osbuild allow the [customization.services] to enable/disable the new units?

Yes, this works and it is already used in the ostree test case which is part of this PR. In short the test does this:

1. Create `/etc/custom_file.txt` with known content.
2. Create `/etc/systemd/system/custom.service` with `ExecStart=/usr/bin/false`.
3. Create a directory `/etc/systemd/system/custom.service.d/` and an `override.conf` inside it with `ExecStart=/usr/bin/cat /etc/custom_file.txt`.
4. Enable `custom.service` using BP Services customization.
5. Check that the `custom.service` is enabled on the booted system, that it was overridden by the drop-in and that the content of `/etc/custom_file.txt` is logged in the journal since the last boot.

> 2. We're pasting in plain text in between quotes. I haven't gotten my hands on this, nor am I a TOML person, so apologies if this is a dumb question. Will users have to escape quotes that are contained inside the pasted text or anything like that?

No need to apologize, I think that there are very few "TOML people" if any 😅 

[TOML allows multiple ways to specify strings](https://toml.io/en/v1.0.0#string) and some of them would require to escape quotes. All of these should be equivalent:

```toml
[[customizations.files]]
path = "/etc/file.txt"
data = "key=\"value\"\n"
```

```toml
[[customizations.files]]
path = "/etc/file.txt"
data = 'key="value"\n'
```

```toml
[[customizations.files]]
path = "/etc/file.txt"
data = """key="value"
"""
```

ondrejbudai reviewed (commented):
This is pure awesomeness! Some small suggestions inline.

thozza commented on internal/fsnode/fsnode.go:
```diff
@@ -0,0 +1,132 @@
+	// Check that the user and group are valid
+	switch user := f.user.(type) {
+	case string:
+		nameRegex := regexp.MustCompile(usernameRegex)
```
Since the code is covered by unit tests, I think that it should be fine to keep it as is...

thozza commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,284 @@
+		return fmt.Errorf("UnmarshalTOML: mode must be a string")
+	}
+
+	switch ensure_parents := dataMap["ensure_parents"].(type) {
```
Ah yes, you are definitely right.... This is the Golang naming convention...

thozza commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,284 @@
+	type fileCustomization FileCustomization
+
+	var filePrivate fileCustomization
+	if err := json.Unmarshal(data, &filePrivate); err != nil {
```
Honestly, I dream of us moving to a more sane TOML library one cay and replacing the `UnmarshalTOML` with something more like `UnmarshalJSON`. The current way TOML unmarshaling way is IMO crazy.

Also it feels cleaner to me to not use TOML unmarshaler inside JSON one...

thozza commented on internal/cloudapi/v2/handler.go:
```diff
@@ -187,6 +187,106 @@ func (h *apiHandlers) PostCompose(ctx echo.Context) error {
 		}
 	}
 
+	if request.Customizations != nil && request.Customizations.Directories != nil {
```
yeah, that is a leftover, good catch!

thozza replied:
Fixed

thozza commented:
I made changes based on reviews. Rebased on top of `main` and also dropped 4 commits, which were related to deploy-time ostree image types and were leftovers from the times when the customization was aiming for them (it is not any more, it instead targets ostree commit image types).

achilleas-k commented on tools/test-case-generators/format-request-map.json:
```diff
@@ -170,114 +170,6 @@
-    },
-    "no-image-info": true,
-    "overrides": {}
-  },
```
:tada: :fishsticks:

achilleas-k commented on test/cases/api.sh:
```diff
@@ -330,6 +330,43 @@ EndOfMessage
+# in every image-type specific file.
+case "${IMAGE_TYPE}" in
+  # The Directories and Files customization is not supported for these image types.
+  "$IMAGE_TYPE_EDGE_COMMIT"|"$IMAGE_TYPE_IOT_COMMIT"|"$IMAGE_TYPE_EDGE_INSTALLER"|"$IMAGE_TYPE_IMAGE_INSTALLER")
```
Didn't we decide we're supporting commits but not `edge-installer`, `edge-raw-image`, and `edge-simplified-installer`?
I see we're testing it in the ostree tests, so is this just for the API?

Also the image installer should be okay to support.  But maybe we can add that later separately.

thozza replied:
> Didn't we decide we're supporting commits but not `edge-installer`, `edge-raw-image`, and `edge-simplified-installer`?

Yes.

>  I see we're testing it in the ostree tests, so is this just for the API?

So I ended up not focusing on testing the commit image in the API test, since it is tested in its own test case. But thinking about this, it should be possible to inspect the custom files in the commit using `ostree`.

Also there is probably no good reason why should I not at least enable the customization for the commit image type.

> Also the image installer should be okay to support. But maybe we can add that later separately.

I already have the support for the `raw-image` and `simplified-installer` implemented on another branch. I removed all of it from this PR, because the decision is to go with just the commit and container image types for 8.8 / 9.2 which this PR targets.

thozza replied:
I added the customization and test for ostree commit in API test.

achilleas-k replied:
> > Also the image installer should be okay to support. But maybe we can add that later separately.

> I already have the support for the raw-image and simplified-installer implemented on another branch. I removed all of it from this PR, because the decision is to go with just the commit and container image types for 8.8 / 9.2 which this PR targets.

I mean the non-ostree Anaconda installer that installs a traditional system (`image-installer`).

achilleas-k replied:
We should instead test it in https://github.com/osbuild/osbuild-composer/blob/main/test/cases/installers.sh in a follow-up.

achilleas-k reviewed (commented):
Looks great!  I have one question but I think everything else is good to go.

thozza commented:
CI Failures:

- ostree / F37 - not caused by this PR, will be fixed by https://github.com/osbuild/osbuild-composer/pull/3292
- vsphere - DNS issues 🤷

supakeen commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,468 @@
+	// Owner of the directory specified as a string (group name), int64 (UID) or nil
+	Group interface{} `json:"group,omitempty" toml:"group,omitempty"`
+	// Permissions of the directory specified as an octal number
+	Mode string `json:"mode,omitempty" toml:"mode,omitempty"`
```
The `User` and `Group` are optional, but not the mode?

thozza replied:
Not sure what do you mean. `mode` is optional. The default value for a string (empty string) is considered to be an unset value. I could have used pointer to string, but there would be no real benefit of it, and at the same time the use would get harder when using string literals (e.g. in unit tests).

supakeen replied:
Understood. It's a Go-ism :)

supakeen commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,468 @@
+	// Permissions of the file specified as an octal number
+	Mode string `json:"mode,omitempty" toml:"mode,omitempty"`
+	// Data is the file content in plain text
+	Data string `json:"data,omitempty" toml:"data,omitempty"`
```
I would prefer to be able to have both `DataAscii` and `DataBytes` to be able to add binary files, we can represent the latter as one of the base encodings in the json/toml?

thozza replied:
I'm with you on this, but the requirement is to support only plain text data for MVP.

As I mentioned in the description, adding support for base64 encoded (potentially binary) data would be super simple, since internaly it's all bytes. My plan would be to later add `data_base64` property later on to hold these. Since the expectation is that users will more often use the "plain text" input, I didn't add "Ascii" into the property name. There is nothing blocking us from renaming the Structure member name to `DataAscii` once the support for base64 encoded content is added, but I didn't see value in doing so now.

achilleas-k replied:
Might also make sense to keep just `Data` and have a toggle for `Binary`?  Or an enum encoding/format, but probably not as simple.  A boolean for `Binary = true/false` would make it easier to handle a single data field without needing to check or tell users that they can't use both.

supakeen commented on internal/blueprint/fsnode_customizations.go:
```diff
@@ -0,0 +1,468 @@
+	// Sort the paths so that we always check the longest paths first. This
+	// ensures that we don't check a parent path before we check the child
+	// path.
+	sort.Sort(sort.Reverse(sort.StringSlice(nodesPaths)))
```
I might be overlooking something but does this sort on strlen of each path? If it does it should sort on dirsep-count.

thozza replied:
You are probably onto something here... 🤔 While the result will be IMO still correct, the optimization does not work as I intended... Will think about this a bit.

achilleas-k replied:
I think I get what your'e saying but I don't see how it can cause an issue.  We only really care that paths are sorted so child paths are ahead of their own parent paths, which this should do.  A path with fewer components being ahead of a path with more components shouldn't cause issues if the former isn't a parent of the latter, right?

thozza replied:
Yeah, there is no impact on the result being correct, but the logic is not correct. I'll just soft the slice using a custom comparison function which will use length based on dir separator.

thozza commented:
- Rebased on `main` to get the F37 fix for ostree test.
- Fixed path sorting to be based on dir separator count in `ValidateDirFileCustomizations()` in the `blueprint` package.
- Added workaround for 💩 DNS making VSphere tests to fail.
- Added dependency on the latest osbuild release (v80) and made use of the `remove_destination` option in the copy stage to not follow symlinks for files customization.
- Pinned osbuild v80 in Schutzfile, since the release happened only yesterday and there was not enough time for it to get to our RPMRepo snapshots and also stable / nightly repos of respective distributions.

achilleas-k reviewed (dismissed):
LGTM!  Great work!  :rocket:

supakeen reviewed (dismissed):
Nice.

thozza commented:
The VSphere workaround has no effect on the test failure and I'm out of ideas, since all three DNS resolvers in `/etc/resolv.conf` are unstable... 😕

thozza commented:
I dropped the commit with "DNS workaround", since it had no effect. IT is working on fixing the DNS resolver in `us-east-1`.
